risc0/bigint2/src/ec/mod.rs

```diff
@@ -16,13 +16,124 @@
 mod tests;
 
 use include_bytes_aligned::include_bytes_aligned;
+use num_bigint::BigUint;
 
-use crate::ffi::sys_bigint2_2;
+use crate::ffi::{sys_bigint2_2, sys_bigint2_3};
 
+const ADD_BLOB: &[u8] = include_bytes_aligned!(4, "add.blob");
 const DOUBLE_BLOB: &[u8] = include_bytes_aligned!(4, "double.blob");
 
-pub fn double(point: &[u32], result: &mut [u32]) {
+pub const EC_256_WIDTH_WORDS: usize = 256 / 32;
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct AffinePoint {
+    /// x coordinate
+    pub x: BigUint,
+    /// y coordinate
+    pub y: BigUint,
+}
+
+impl AffinePoint {
+    /// The point as concatenated u32s for x and y
+    ///
+    /// Little-endian, x coordinate before y coordinate
+    #[cfg(not(feature = "num-bigint-dig"))]
+    pub fn to_u32s(&self) -> [u32; 2 * EC_256_WIDTH_WORDS] {
+        // TODO: This feels duplicative with `to_u32_digits` from RSA, but I don't see a way to share code without doubling the `copy_from_slice` calls
+        let mut result = [0u32; 2 * EC_256_WIDTH_WORDS];
+        let mut first = self.x.to_u32_digits();
+        assert!(first.len() <= EC_256_WIDTH_WORDS);
+        if first.len() < EC_256_WIDTH_WORDS {
+            first.resize(EC_256_WIDTH_WORDS, 0);
+        }
+        let mut last = self.y.to_u32_digits();
+        assert!(last.len() <= EC_256_WIDTH_WORDS);
+        if last.len() < EC_256_WIDTH_WORDS {
+            last.resize(EC_256_WIDTH_WORDS, 0);
+        }
```

Comment:
We should try to avoid resizes

Reply:
:+1: I think I can drop these in the generalized EC architecture, I'll make sure to pay attention to that